src/typecheck/mod.rs

```diff
@@ -1077,9 +1085,41 @@ fn walk_with_annot<L: Linearizer>(
     }
 }
 
-/// Typecheck a term against a specific type.
+/// Check a term against a given type. Although this method mostly corresponds to checking mode in
+/// the classical bidirectional framework, it combines both checking and inference mode in
+/// practice, to avoid duplicating rules (that is, code) as detailed below.
+///
+/// # Literals
+///
+/// Checking a literal (a number, a string, a boolean, etc.) unifies the checked type with the
+/// corresponding primitive type (`Number`, `String`, `Bool`, etc.). If the checked type is a
+/// unification variable, `check` acts as an inference rule. If the type is concrete, unification
+/// enforces equality, and `check` acts as a checking rule.
+///
+/// # Introduction rules
 ///
-/// # Arguments
+/// Following Pfenning's recipe (see [Bidirectional Typing][bidirectional-typing]),
+/// introduction rules (e.g. typechecking a record) are checking. `check` follows the same logic
+/// here: it uses unification to "match" on the expected type (in the case of record, a record type
+/// or a dictionary type) and pushes typechecking down the record fields.
+///
+/// # Elimination rules
+///
+/// Elimination rules (such as function application or primitive operator application) are only in
+/// inference mode (still following Pfenning's recipe). `check` follows the inference mode here
+/// (typically on function application, where we first call to `infer` on the function part, and
+/// then check the argument).
+///
+/// Still, `check` is supposed to be implementing checking mode from the outside. We thus also
+/// apply the rule which switches from inference to checking mode. This is in particular where
+/// subtyping happens, for example, when there is subtyping. Currently, subtyping isn't supported
```

Comment:
```
This is in particular where subtyping happens, for example, when there is subtyping.
```

I'm not entirely sure what this sentence is trying to say. :sweat_smile:

Reply:
:sweat_smile: the switch between inference and checking mode is where we apply subtyping in the general framework. Currently, it's a degenerate case, because there is no subtyping yet (or, viewed differently, subtyping is trivially the equality relation). But it's indeed quite confusing. I think @vkleen suggestion below makes it better.

Reply:
Tentative improvement: https://github.com/tweag/nickel/pull/1193/commits/3b674c88b6c2187cdce6eca9e79f74153d789e1b

Reply:
I like it 👍